src/entities/addLiquidity/addLiquidityCowAmm/index.ts

```diff
@@ -0,0 +1,135 @@
+import {
+    createPublicClient,
+    encodeFunctionData,
+    formatEther,
+    formatUnits,
+    http,
+} from 'viem';
+import { cowAmmPoolAbi } from '@/abi/cowAmmPool';
+import { HumanAmount } from '@/data';
+import { Token } from '@/entities/token';
+import { TokenAmount } from '@/entities/tokenAmount';
+import { PoolState, PoolStateWithBalances } from '@/entities/types';
+import {
+    calculateProportionalAmounts,
+    getSortedTokens,
+    getValue,
+} from '@/entities/utils';
+import { CHAINS } from '@/utils';
+
+import { getAmountsCall } from '../helpers';
+import {
+    AddLiquidityBase,
+    AddLiquidityBaseBuildCallInput,
+    AddLiquidityBaseQueryOutput,
+    AddLiquidityBuildCallOutput,
+    AddLiquidityKind,
+    AddLiquidityProportionalInput,
+} from '../types';
+import {
+    getPoolTokenBalanceCowAmm,
+    getTotalSupplyCowAmm,
+} from '../../utils/cowAmmHelpers';
+
+export class AddLiquidityCowAmm implements AddLiquidityBase {
+    async query(
+        input: AddLiquidityProportionalInput,
+        poolState: PoolState,
+    ): Promise<AddLiquidityBaseQueryOutput> {
+        const sortedTokens = getSortedTokens(poolState.tokens, input.chainId);
+
+        const client = createPublicClient({
+            transport: http(input.rpcUrl),
+            chain: CHAINS[input.chainId],
+        });
+
+        const balances: bigint[] = [];
+        for (const token of sortedTokens) {
+            balances.push(
+                await getPoolTokenBalanceCowAmm(
```

Comment:
Instead of doing multiple separate RPC calls can we do a [multicall](https://viem.sh/docs/contract/multicall) (along with totalSupply) to get all the info in a single call?

Reply:
done ✅